service/src/main/java/io/camunda/service/AdHocSubprocessActivityServices.java

```diff
@@ -0,0 +1,115 @@
+      final Authentication authentication) {
+    super(brokerClient, securityContextProvider, authentication);
+    this.processDefinitionServices = processDefinitionServices;
+  }
+
+  @Override
+  public AdHocSubprocessActivityServices withAuthentication(final Authentication authentication) {
+    return new AdHocSubprocessActivityServices(
+        brokerClient,
+        securityContextProvider,
+        processDefinitionServices.withAuthentication(authentication),
+        authentication);
+  }
+
+  public List<AdHocSubprocessActivity> findActivatableActivities(
+      final Long processDefinitionKey, final String adHocSubprocessId) {
+    final ProcessDefinitionEntity processDefinitionEntity =
+        processDefinitionServices.getByKey(processDefinitionKey);
+
+    // TODO should we cache this lookup somewhere? There are multiple ProcessCache implementations
+    // in operate, tasklist, gateway-rest, but none in this service layer. Implement caching later
+    // or introduce a process definition cache at this level?
+    final BpmnModelInstance modelInstance =
+        Bpmn.readModelFromStream(
+            new ByteArrayInputStream(
+                processDefinitionEntity.bpmnXml().getBytes(StandardCharsets.UTF_8)));
+
+    final var processElement = modelInstance.getModelElementById(adHocSubprocessId);
+    if (processElement instanceof final AdHocSubProcess adHocSubProcess) {
+      final var tasks = adHocSubProcess.getChildElementsByType(Task.class);
+      return tasks.stream()
+          .filter(task -> task.getIncoming().isEmpty())
+          .map(
+              task ->
+                  new AdHocSubprocessActivity(
+                      processDefinitionKey,
+                      adHocSubprocessId,
+                      task.getId(),
+                      task.getName(),
+                      BpmnElementType.bpmnElementTypeFor(task.getElementType().getTypeName()),
+                      extractDocumentation(task)))
+          .toList();
+    }
+
+    if (LOGGER.isWarnEnabled()) {
+      if (processElement == null) {
+        LOGGER.warn(
+            "Failed to find ad-hoc subprocess element with id '{}' in process definition with key '{}'",
+            adHocSubprocessId,
+            processDefinitionKey);
+      } else {
+        LOGGER.warn(
+            "Found element with id '{}' in process definition with key '{}' but it was of type '{}', expected 'AdHocSubProcess'",
+            adHocSubprocessId,
+            processDefinitionKey,
+            processElement.getElementType());
+      }
+    }
+
+    throw new NotFoundException(
+        "Failed to find Ad-Hoc Subprocess with ID '%s'".formatted(adHocSubprocessId));
+  }
+
+  private String extractDocumentation(final Task task) {
+    return task.getDocumentations().stream()
+        .map(ModelElementInstance::getTextContent)
+        .collect(Collectors.joining("\n"));
+  }
+
+  public record AdHocSubprocessActivity(
```

Comment:
I did not (yet) find a good package for these data objects for use cases not related to search so for the moment I just put this here also if it does not 100% fit how other services look like. Open to discuss :)